command/job_restart.go

```diff
@@ -0,0 +1,850 @@
+	for _, task := range c.tasks.Slice() {
+		if _, ok := alloc.TaskStates[task]; !ok {
+			continue
+		}
+
+		c.ui.Output(fmt.Sprintf(
+			"    %s: Restarting task %q in allocation %q for group %q",
+			formatTime(time.Now()),
+			task,
+			shortAllocID,
+			alloc.TaskGroup,
+		))
+
+		restarts.Go(func(task string) func() error {
+			return func() error {
+				return c.client.Allocations().Restart(alloc, task, nil)
+			}
+		}(task))
+	}
+	return restarts.Wait().ErrorOrNil()
+}
+
+// stopAlloc stops an allocation and blocks until the replacement allocation is
+// running.
+func (c *JobRestartCommand) stopAlloc(alloc *api.Allocation) error {
+	shortAllocID := limit(alloc.ID, c.length)
+
+	c.ui.Output(fmt.Sprintf(
+		"    %s: Rescheduling allocation %q for group %q",
+		formatTime(time.Now()),
+		shortAllocID,
+		alloc.TaskGroup,
+	))
+
+	var q *api.QueryOptions
+	if c.noShutdownDelay {
+		q = &api.QueryOptions{
+			Params: map[string]string{"no_shutdown_delay": "true"},
+		}
+	}
+
+	// Stop allocation and use a blocking query to wait for the evaluation to
+	// be processed by the scheduler.
+	resp, err := c.client.Allocations().Stop(alloc, q)
+	if err != nil {
+		return fmt.Errorf("Error stopping allocation %s: %v", shortAllocID, err)
+	}
+
+	q = &api.QueryOptions{WaitIndex: 1}
+	var qm *api.QueryMeta
+	var eval *api.Evaluation
+	for {
+		eval, qm, err = c.client.Evaluations().Info(resp.EvalID, q)
+		if err != nil {
+			return fmt.Errorf("Error retrieving evaluation %s: %v", resp.EvalID, err)
+		}
+		if eval.Status != api.EvalStatusPending {
+			break
+		}
+		q.WaitIndex = qm.LastIndex
+	}
+
+	// Check for placement failures.
+	var mErr *multierror.Error
+	for tg, metrics := range eval.FailedTGAllocs {
+		failures := metrics.CoalescedFailures + 1
+		mErr = multierror.Append(mErr, fmt.Errorf(
+			"Task Group %q (failed to place %s):\n%s",
+			tg,
+			english.Plural(failures, "allocation", "allocations"),
+			formatAllocMetrics(metrics, false, strings.Repeat(" ", 4)),
+		))
+	}
+	if err := mErr.ErrorOrNil(); err != nil {
+		return err
+	}
+
+	// Use a blocking query to re-fetch the allocation that was just stopped to
+	// find the follow-up allocation.
+	q = &api.QueryOptions{WaitIndex: 1}
+	var nextAllocID string
+	for {
+		oldAlloc, qm, err := c.client.Allocations().Info(alloc.ID, q)
+		if err != nil {
+			return fmt.Errorf("Error retrieving allocation %s: %v", shortAllocID, err)
+		}
+		if oldAlloc.NextAllocation != "" {
+			nextAllocID = oldAlloc.NextAllocation
+			break
+		}
+		q.WaitIndex = qm.LastIndex
+	}
+
+	// Use a blocking query to wait for the follow-up allocation to be running.
+	q = &api.QueryOptions{WaitIndex: 1}
+	for {
+		newAlloc, qm, err := c.client.Allocations().Info(nextAllocID, q)
+		if err != nil {
+			return fmt.Errorf("Error retrieving replacement allocation %s: %v", limit(nextAllocID, c.length), err)
+		}
+		if newAlloc.ClientStatus == api.AllocClientStatusRunning {
+			break
+		}
+		q.WaitIndex = qm.LastIndex
+	}
```

Comment:
Would it be worth to have a flag that skips this check for a faster restart? Or maybe something like `-wait-for-status=pending` so it doesn't always wait for `running`, which could take a while for some workloads.

Reply:
Oh that's a good point. Not waiting for a successful outcome can only lead to bad things 😅

Reply:
If the followup allocation terminates before this catches it running it will loop forever (well... until the alloc is GC'd and `Info(...)` returns error). Even watching the event stream and processing every update for the alloc isn't enough as clients and servers batch updates, so I think it's possible for an alloc to go directly from `pending -> failed`.

Seems like if we error on `ClientStatusTerminal()` with a recoverable error, the batch will be allowed to proceed unless `-fail-on-error` is set? That seems reasonable?

For allocs that are stuck in pending or can't be schedule (the FailedTGAllocs check above won't failures like an eval being canceled or failing), I'm not sure what to do. This will block forever in a fairly silent and mysterious way.

Perhaps we could slightly change `-batch-wait` to be more like `-batch-duration` and be the time the command allots to performing a batch of operations. These polling queries could log to inform the user if they want to stop before the next batch.

I'm not sure what the right answer here is, but I am worried trying this path only works without edge cases if you effectively recreate deployments in the CLI... at which point we should just create a helper for the `bump a meta{} var` trick.

Reply:
One idea on error handling. Instead of having `-fail-on-error` have a flag `-on-error` that can be set to :
* `ask` (default for TTY sessions) behaving like `-batch-wait` and asking the user how to proceed if an error happens during a batch.
* `fail` that behaves like the current `-fail-on-error`
* `continue` (default for non-TTY?) that behaves like the current default and accumulates errors until the command exits.

Reply:
> > For allocs that are stuck in pending or can't be schedule (the FailedTGAllocs check above won't failures like an eval being canceled or failing), I'm not sure what to do. This will block forever in a fairly silent and mysterious way.
>
> Is there an easy way to simulate this? All my tests had FailedTGAllocs populated thinking

Canceled occurs when multiple evals are enqueued for the same job. Probably only occurs under scheduler load and/or when concurrent cluster activity causes the job to be eval'd. I think it could be exercised by setting NumSchedulers=0, kicking off a `restart -reschedule job-with-multiple-allocs`, and then setting NumSchedulers=1.

https://github.com/hashicorp/nomad/blob/b07af5761846a59ac91e4826c9a0f8e38c91f70b/nomad/eval_broker.go#L594-L610

I think failed evals are even weirder and rarer edge cases that occur when an eval has been nack'd too many times. Nacks occur when the worker/scheduler hits an error/timeout. Not sure that's worth trying to test. :grimacing: 

https://github.com/hashicorp/nomad/blob/b07af5761846a59ac91e4826c9a0f8e38c91f70b/nomad/worker.go#L417-L423


> One idea on error handling. Instead of having -fail-on-error have a flag -on-error that can be set to :

I think this works. `continue` is pretty scary as it allows for Nomad killing your job while you're not paying attention. Maybe the non-TTY default should be `fail` and require users to opt into the scary option?

Reply:
> I think it could be exercised by setting NumSchedulers=0, kicking off a restart -reschedule job-with-multiple-allocs, and then setting NumSchedulers=1.

Oh nice, I will try that next week. Thanks!

> `continue` is pretty scary as it allows for Nomad killing your job while you're not paying attention. Maybe the non-TTY default should be `fail` and require users to opt into the scary option?

Good point. I added a new [`-yes` flag](https://github.com/hashicorp/nomad/pull/16278/files#diff-7a75798ae34efc4d2d03c5950d047c4d5efe878631f5a43712ddcecbde45c696R99) to handle multi-region jobs and I think it can be useful here as well.

`ask` will be the default but fail validation on non-TTY session like `-batch-wait`. To always proceed even in case of errors you will need to add a `-yes`.